pkg/controllers/netpol/policy.go

```diff
@@ -79,72 +88,172 @@ func (npc *NetworkPolicyController) syncNetworkPolicyChains(networkPoliciesInfo
+		}
+		err = npc.processIngressRules(policy, targetDestPodIPSetName, activePolicyIPSets, version)
+		if err != nil {
+			return err
+		}
+		activePolicyIPSets[targetDestPodIPSet.Name] = true
+	}
 
-		currnetPodIps := make([]string, 0, len(policy.targetPods))
-		for ip := range policy.targetPods {
-			currnetPodIps = append(currnetPodIps, ip)
+	if policy.policyType == "both" || policy.policyType == "egress" {
+		// create a ipset for all source pod ip's matched by the policy spec PodSelector
+		targetSourcePodIPSetName := policySourcePodIPSetName(policy.namespace, policy.name)
+		targetSourcePodIPSet, err := npc.ipSetHandler.Create(targetSourcePodIPSetName, utils.TypeHashIP, utils.OptionTimeout, "0")
+		if err != nil {
+			return fmt.Errorf("failed to create ipset: %s", err.Error())
+		}
+		err = targetSourcePodIPSet.Refresh(currnetPodIps, utils.OptionTimeout, "0")
+		if err != nil {
+			glog.Errorf("failed to refresh targetSourcePodIPSet: " + err.Error())
 		}
+		err = npc.processEgressRules(policy, targetSourcePodIPSetName, activePolicyIPSets, version)
+		if err != nil {
+			return err
+		}
+		activePolicyIPSets[targetSourcePodIPSet.Name] = true
+	}
+	return nil
+}
 
-		if policy.policyType == "both" || policy.policyType == "ingress" {
-			// create a ipset for all destination pod ip's matched by the policy spec PodSelector
-			targetDestPodIPSetName := policyDestinationPodIPSetName(policy.namespace, policy.name)
-			targetDestPodIPSet, err := npc.ipSetHandler.Create(targetDestPodIPSetName, utils.TypeHashIP, utils.OptionTimeout, "0")
-			if err != nil {
-				return nil, nil, fmt.Errorf("failed to create ipset: %s", err.Error())
-			}
-			err = targetDestPodIPSet.Refresh(currnetPodIps, utils.OptionTimeout, "0")
-			if err != nil {
-				glog.Errorf("failed to refresh targetDestPodIPSet,: " + err.Error())
+// when a new pod added/deleted this function ensures only matching network
+// policies (i.e. pod labels match with network policy target pod selector or ingress pod
+// selector or egress pod selector) are synced to reflect the desired state
+func (npc *NetworkPolicyController) syncAffectedNetworkPolicyChains(
+	pod *podInfo,
+	version string) error {
+
+	activePolicyChains := make(map[string]bool)
+	activePolicyIPSets := make(map[string]bool)
+
+	for _, policyObj := range npc.npLister.List() {
+		var matchesTargetPod, matchesIngressSrcPod, matchesEgressDestPod bool
+		policy, ok := policyObj.(*networking.NetworkPolicy)
+		targetPodSelector, _ := v1.LabelSelectorAsSelector(&policy.Spec.PodSelector)
+		if !ok {
+			return fmt.Errorf("Failed to convert")
+		}
+		if policy.ObjectMeta.Namespace == pod.namespace && targetPodSelector.Matches(labels.Set(pod.labels)) {
+			matchesTargetPod = true
+		}
+		for _, specIngressRule := range policy.Spec.Ingress {
+			if len(specIngressRule.From) == 0 {
+				continue
+			}
+			for _, peer := range specIngressRule.From {
+				var matchesPodSelector, matchesNamespaceSelector bool
```

Comment:
This doesn't take into account CIDR network policies which will also have nil PodSelector or NameSpaceSelectors.

Something like the following needs to be added to Ingress and Egress rule processing so as to not try to unnecessarily process CIDR policy changes
```
				if peer.IPBlock != nil {
					continue
				}
```